crates/extensions/tedge_flows/src/connected_flow.rs

```diff
@@ -0,0 +1,180 @@
+use crate::config::ConfigError;
+use crate::flow::DateTime;
+use crate::flow::Flow;
+use crate::flow::FlowError;
+use crate::flow::FlowInput;
+use crate::flow::FlowResult;
+use crate::flow::Message;
+use crate::input_source::CommandPollingSource;
+use crate::input_source::CommandStreamingSource;
+use crate::input_source::FilePollingSource;
+use crate::input_source::FileStreamingSource;
+use crate::input_source::PollingSource;
+use crate::input_source::StreamingSource;
+use crate::registry::FlowRegistry;
+use crate::registry::FlowStore;
+use camino::Utf8Path;
+use tedge_watch_ext::WatchRequest;
+use tokio::time::Instant;
+
+/// A flow connected to a source of messages
+pub struct ConnectedFlow {
+    flow: Flow,
+    streaming_source: Option<Box<dyn StreamingSource>>,
+    polling_source: Option<Box<dyn PollingSource>>,
+}
+
+impl AsRef<Flow> for ConnectedFlow {
+    fn as_ref(&self) -> &Flow {
+        &self.flow
+    }
+}
+
+impl AsMut<Flow> for ConnectedFlow {
+    fn as_mut(&mut self) -> &mut Flow {
+        &mut self.flow
+    }
+}
+
+impl ConnectedFlow {
+    pub fn new(flow: Flow) -> Self {
+        let streaming_source = streaming_source(flow.name().to_owned(), flow.input.clone());
+        let polling_source = polling_source(flow.input.clone());
+        ConnectedFlow {
+            flow,
+            streaming_source,
+            polling_source,
+        }
+    }
+
+    pub fn name(&self) -> &str {
+        self.flow.name()
+    }
+
+    pub fn input_topic(&self) -> &str {
+        self.flow.input.enforced_topic().unwrap_or_default()
+    }
+
+    pub fn watch_request(&self) -> Option<WatchRequest> {
+        self.streaming_source
+            .as_ref()
+            .and_then(|source| source.watch_request())
+    }
+
+    pub fn next_deadline(&self) -> Option<Instant> {
+        self.polling_source.as_ref().and_then(|p| p.next_deadline())
+    }
+
+    pub async fn on_source_poll(&mut self, timestamp: DateTime, now: Instant) -> FlowResult {
+        let result = self.on_source_poll_steps(timestamp, now).await;
+        self.flow.publish(result)
+    }
+
+    async fn on_source_poll_steps(
+        &mut self,
+        timestamp: DateTime,
+        now: Instant,
+    ) -> Result<Vec<Message>, FlowError> {
+        let Some(source) = &mut self.polling_source.as_mut() else {
+            return Ok(vec![]);
+        };
+        if !source.is_ready(now) {
+            return Ok(vec![]);
+        };
+
+        let messages = source.poll(timestamp).await?;
+        source.update_after_poll(now);
+        Ok(messages)
+    }
+
+    pub fn on_error(&self, error: FlowError) -> FlowResult {
+        self.flow.publish(Err(error))
+    }
+}
+
+fn streaming_source(flow_name: String, input: FlowInput) -> Option<Box<dyn StreamingSource>> {
+    match input {
+        FlowInput::PollFile {
+            topic: _,
+            path,
+            interval,
+        } if interval.is_zero() => Some(Box::new(FileStreamingSource::new(flow_name, path))),
```

Comment:
I'm slightly suprised to see this logic here rather than when we generate the `FlowInput`, particularly as it's reliant on `streaming_source` and `polling_source` being in agreement about what we need to do to handle this case.